In [13]:
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
import numpy as np

## 피쳐 전처리

In [14]:
from sklearn.preprocessing import StandardScaler

scaler = StandardScaler()

def remove_outlier(data_df):
    data_df = data_df.copy()
    data_df = data_df.drop(data_df[data_df["minimum_nights"] > 400].index)
    data_df = data_df.drop(data_df[data_df["price"] > 5000].index)
    return data_df

def preprocess_data(X, is_train):
    X = X.copy()

    # 사용하지 않는 피쳐 제거
    X = X.drop(columns=[
        "id",
        "name",
        "host_id",
        "host_name",
        "calculated_host_listings_count",
        "neighbourhood",
        "availability_365",
        "last_review",
    ])

    # neighbourhood_group: one-hot encoding
    X = pd.get_dummies(X, columns=["neighbourhood_group"], drop_first=True)

    # room_type one-hot encoding
    X = pd.get_dummies(X, columns=["room_type"], drop_first=True)

    # mimimum_nights: log 변환
    X["minimum_nights"] = np.log1p(X["minimum_nights"])

    # number_of_reviews, reviews_per_month: 차원 축소 및 log 변환
    X = X.drop(columns=["reviews_per_month"])
    X["number_of_reviews"] = np.log1p(X["number_of_reviews"])

    # feature scaling
    numeric_features = X.select_dtypes("number").columns.tolist()
    if is_train:
        scaler.fit(X[numeric_features])
    X[numeric_features] = scaler.transform(X[numeric_features])

    return X

In [15]:
train_df = pd.read_csv("./data/train.csv")
test_df = pd.read_csv("./data/test.csv")
y_test_df = pd.read_csv("./data/y_test.csv")

# train
train_df = remove_outlier(train_df)
X_train = train_df.drop(columns=["price"])
X_train = preprocess_data(X_train, is_train=True)
y_train = train_df["price"]

# test
X_test = test_df.copy()
X_test = preprocess_data(X_test, is_train=False)
y_test = y_test_df.copy()["price"]

X_train.shape, y_train.shape, X_test.shape, y_test.shape

((39092, 10), (39092,), (9779, 10), (9779,))

## baseline model과 비교

In [16]:
from xgboost import XGBRegressor
from sklearn.metrics import mean_squared_error

model = XGBRegressor(random_state=2022)

model.fit(X_train, y_train)
pred = model.predict(X_test)
mean_squared_error(y_test, pred)

48387.558380601826

## Model 최적화

In [17]:
from pycaret.regression import *

reg = setup(data=X_train, target=y_train,
            train_size=0.8, fold=5, 
            n_jobs=-1,
            session_id=42)

best_models = compare_models(sort="mse", n_select=3)
best_models

[LGBMRegressor(n_jobs=-1, random_state=42),
 GradientBoostingRegressor(random_state=42),
 Ridge(random_state=42)]

In [18]:
best_tuned_models = [tune_model(m) for m in best_models]
blended_model = blend_models(best_tuned_models)
model = finalize_model(blended_model)

Initiated  . . . . . . . . . . . . . . . . . .              16:22:28
Status     . . . . . . . . . . . . . . . . . .  Loading Dependencies
Estimator  . . . . . . . . . . . . . . . . . .     Compiling Library

Fitting 5 folds for each of 10 candidates, totalling 50 fits


Fitting 5 folds for each of 10 candidates, totalling 50 fits


Fitting 5 folds for each of 10 candidates, totalling 50 fits
Original model was better than the tuned model, hence it will be returned. NOTE: The display metrics are for the tuned model (not the original one).


In [19]:
model

Pipeline(memory=Memory(location=None),
         steps=[('numerical_imputer',
                 TransformerWrapper(include=['latitude', 'longitude',
                                             'minimum_nights',
                                             'number_of_reviews'],
                                    transformer=SimpleImputer())),
                ('categorical_imputer',
                 TransformerWrapper(include=[],
                                    transformer=SimpleImputer(strategy='most_frequent'))),
                ('clean_column_names',
                 TransformerWrapper(transformer=Clean...
                                                            n_jobs=-1,
                                                            num_leaves=40,
                                                            random_state=42,
                                                            reg_alpha=2,
                                                            reg_lambda=0.001)),
                                             ('Gradient Boosting Regressor',
                                              GradientBoostingRegressor(learning_rate=0.05,
                                                                        max_depth=6,
                                                                        max_features='sqrt',
                                                                        min_impurity_decrease=0.0002,
                                                                        min_samples_leaf=3,
                                                                        min_samples_split=5,
                                                                        n_estimators=120,
                                                                        random_state=42,
                                                                        subsample=0.65)),
                                             ('Ridge Regression',
                                              Ridge(random_state=42))],
                                 n_jobs=-1))])

In [20]:
from sklearn.metrics import mean_squared_error

y_pred = predict_model(model, data=X_train)["prediction_label"]
print("Train data score:", mean_squared_error(y_train, y_pred))

y_pred = predict_model(model, data=X_test)["prediction_label"]
print("Test data score:", mean_squared_error(y_test, y_pred))

[LightGBM] [Warning] bagging_freq is set=5, subsample_freq=0 will be ignored. Current value: bagging_freq=5
[LightGBM] [Warning] feature_fraction is set=0.8, colsample_bytree=1.0 will be ignored. Current value: feature_fraction=0.8
[LightGBM] [Warning] bagging_fraction is set=0.6, subsample=1.0 will be ignored. Current value: bagging_fraction=0.6
Train data score: 25328.76009025393


[LightGBM] [Warning] bagging_freq is set=5, subsample_freq=0 will be ignored. Current value: bagging_freq=5
[LightGBM] [Warning] feature_fraction is set=0.8, colsample_bytree=1.0 will be ignored. Current value: feature_fraction=0.8
[LightGBM] [Warning] bagging_fraction is set=0.6, subsample=1.0 will be ignored. Current value: bagging_fraction=0.6
Test data score: 47520.259942760786


---

- Baseline: 54244.32571482477
- Final: 47520.259942760786

과제를 진행하면서 다양한 방법을 적용해보았는데 특히 pycaret을 처음 사용하면서 모델의 최적화를 시도해보았다.  
다만 피쳐 엔지니어링 방법이 적절한지에 대해서는 아직 의문이 든다.  
여러 코드를 보면서 방법에 대해서 비교해봐야할 것 같다.